python/cuda_parallel/benchmarks/bench_merge_sort.py

```diff
@@ -3,59 +3,121 @@
 
 import cuda.parallel.experimental.algorithms as algorithms
 import cuda.parallel.experimental.iterators as iterators
+from cuda.parallel.experimental.struct import gpu_struct
 
 
-def merge_sort_pointer(size, build_only):
-    keys = cp.arange(size, dtype="i4")
-    vals = cp.arange(size, dtype="i8")
-    res_keys = cp.empty_like(keys)
-    res_vals = cp.empty_like(vals)
+def merge_sort_pointer(keys, vals, output_keys, output_vals, build_only):
+    size = len(keys)
 
     def my_cmp(a: np.int32, b: np.int32) -> np.int32:
         return np.int32(a < b)
 
-    alg = algorithms.merge_sort(keys, vals, res_keys, res_vals, my_cmp)
-    temp_bytes = alg(None, keys, vals, res_keys, res_vals, size)
+    alg = algorithms.merge_sort(keys, vals, output_keys, output_vals, my_cmp)
+    temp_bytes = alg(None, keys, vals, output_keys, output_vals, size)
     scratch = cp.empty(temp_bytes, dtype=cp.uint8)
 
     if not build_only:
-        alg(scratch, keys, vals, res_keys, res_vals, size)
+        alg(scratch, keys, vals, output_keys, output_vals, size)
 
     cp.cuda.runtime.deviceSynchronize()
 
 
-def merge_sort_iterator(size, build_only):
+def merge_sort_iterator(size, output_keys, output_vals, build_only):
     keys_dt = cp.int32
     vals_dt = cp.int64
     keys = iterators.CountingIterator(np.int32(0))
     vals = iterators.CountingIterator(np.int64(0))
-    res_keys = cp.empty(size, dtype=keys_dt)
-    res_vals = cp.empty(size, dtype=vals_dt)
+    output_keys = cp.empty(size, dtype=keys_dt)
+    output_vals = cp.empty(size, dtype=vals_dt)
 
     def my_cmp(a: np.int32, b: np.int32) -> np.int32:
         return np.int32(a < b)
 
-    alg = algorithms.merge_sort(keys, vals, res_keys, res_vals, my_cmp)
-    temp_bytes = alg(None, keys, vals, res_keys, res_vals, size)
+    alg = algorithms.merge_sort(keys, vals, output_keys, output_vals, my_cmp)
+    temp_bytes = alg(None, keys, vals, output_keys, output_vals, size)
     scratch = cp.empty(temp_bytes, dtype=cp.uint8)
 
     if not build_only:
-        alg(scratch, keys, vals, res_keys, res_vals, size)
+        alg(scratch, keys, vals, output_keys, output_vals, size)
+
+    cp.cuda.runtime.deviceSynchronize()
+
+
+@gpu_struct
+class MyStruct:
+    x: np.int32
+    y: np.int32
+
+
+def merge_sort_struct(size, keys, vals, output_keys, output_vals, build_only):
+    size = len(keys)
+
+    def my_cmp(a: MyStruct, b: MyStruct) -> np.int8:
+        return np.int8(a.x < b.x)
+
+    alg = algorithms.merge_sort(keys, vals, output_keys, output_vals, my_cmp)
+
+    if not build_only:
+        temp_bytes = alg(None, keys, vals, output_keys, output_vals, size)
+        scratch = cp.empty(temp_bytes, dtype=cp.uint8)
```

Comment:
This is inconsistent with the above benchmarks, where you do not include the time to create temporary storage when running the algorithm.

Reply:
Fixed (ditto below)